helix-term/src/commands.rs

```diff
@@ -5516,14 +5519,41 @@ fn shrink_selection(cx: &mut Context) {
         // if not previous selection, shrink to first child
         if let Some(syntax) = doc.syntax() {
             let text = doc.text().slice(..);
-            let selection = object::shrink_selection(syntax, text, current_selection);
+            let selection = object::shrink_selection(syntax, text, current_selection.clone());
             doc.set_selection(view.id, selection);
         }
     };
     motion(cx.editor);
     cx.editor.last_motion = Some(Motion(Box::new(motion)));
 }
 
+fn select_sibling_impl<F>(cx: &mut Context, sibling_fn: &'static F)
```

Comment:
I think we can remove the `&'static` and just leave `F`.

Reply:
the borrow checker gets upset here without the `&`:

```
error[E0507]: cannot move out of `sibling_fn`, a captured variable in an `Fn` closure
  --> helix-core/src/object.rs:30:44
   |
24 |       sibling_fn: F,
   |       ---------- captured outer variable
...
29 |       select_node_impl(syntax, text, selection, |descendant, _from, _to| {
   |  _______________________________________________-
30 | |         find_sibling_recursive(descendant, sibling_fn)
   | |                                            ^^^^^^^^^^ move occurs because `sibling_fn` has type `F`, which does not implement the `Copy` trait
31 | |     })
   | |_____- captured by this `Fn` closure

For more information about this error, try `rustc --explain E0507`.
```

and also if we only remove the `'static` lifetime bounds:

```
error[E0310]: the parameter type `F` may not live long enough
    --> helix-term/src/commands.rs:5546:41
     |
5530 | fn select_sibling_impl<F>(cx: &mut Context, sibling_fn: &F)
     |                        - help: consider adding an explicit lifetime bound...: `F: 'static`
...
5546 |     cx.editor.last_motion = Some(Motion(Box::new(motion)));
     |                                         ^^^^^^^^ ...so that the reference type `&F` does not outlive the data it points at

error[E0310]: the parameter type `F` may not live long enough
    --> helix-term/src/commands.rs:5546:41
     |
5530 | fn select_sibling_impl<F>(cx: &mut Context, sibling_fn: &F)
     |                        - help: consider adding an explicit lifetime bound...: `F: 'static`
...
5546 |     cx.editor.last_motion = Some(Motion(Box::new(motion)));
     |                                         ^^^^^^^^^^^^^^^^ ...so that the type `[closure@helix-term/src/commands.rs:5534:18: 5544:6]` will meet its required lifetime bounds

For more information about this error, try `rustc --explain E0310`.
```

Is there a way to make the borrow checker happy in these cases?

Reply:
From what I understand the `'static` bound is necessary for functions and there isn't such a thing as a non-`'static` fn. I can't find a decent explainer article for it anymore though